wordpress-mobile/WordPress-iOS
# Reader: Add announcement card

Part of #23069 

Adds the announcement card to the Reader stream. The card is only shown when:

- There's no other header being displayed. The announcement card should have the lowest display priority.
- There are no active filters.
- There are contents to display. It shouldn't be displayed when the no results or error view is displayed.
- The announcement card hasn't been dismissed yet.
- Lastly, of course, the remote feature flag `reader_announcement_card` needs to be enabled.

Preview:

Light | Dark
-|-
![Simulator Screenshot - iPhone 15 - 2024-05-14 at 02 25 27](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/0669564e-81ef-4725-b802-8bf90471e331) | ![Simulator Screenshot - iPhone 15 - 2024-05-14 at 02 20 20](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/6bb7c868-8bea-4a47-9554-a2878e2dbb92)


## To test

> [!TIP]
> To avoid dismissing the card and never seeing it again, you could temporarily disable the dismiss logic by commenting out the line in `ReaderStreamViewController+Helper.swift:L179`:
> ```swift
>     self?.readerAnnouncementCoordinator.isDismissed = false
> ```
> This way, the announcement card will reappear on the next stream after you dismiss it.


- Make sure the `readerAnnouncementCard` flag is enabled.
- Launch the Jetpack app.
- Go to the Reader tab.
- 🔎 Verify that the announcement card is displayed in the Discover stream.
- Navigate to other (non-empty) streams.
- 🔎 Verify that the announcement card is visible.
- Navigate to an empty stream.
- 🔎 Verify that the announcement card is NOT visible.
- Go to `Subscriptions` or `Your Tags` and filter the stream.
- 🔎 Verify that the announcement card is NOT visible.

#### Offline case
- Turn on airplane mode and relaunch the Jetpack app.
- Go to the Reader stream > Discover.
- 🔎 Verify that the announcement card is NOT visible.

#### Spam test

<details>

<summary>Context</summary>

During development, I stumbled on some crashes while switching to `Discover`, and it crashed with this message:

```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'unable to dequeue a cell with identifier ReaderSitesCell - must register a nib or a class for the identifier or connect a prototype cell in a storyboard'
```

https://github.com/wordpress-mobile/WordPress-iOS/blob/c055c21bfb03461fd18bb70f53ba62f25c38a2de/WordPress/Classes/ViewRelated/Reader/ReaderCardsStreamViewController.swift#L93

However, after a while, it seems like the crash stopped happening... I'm guessing it may be related to this commit: https://github.com/wordpress-mobile/WordPress-iOS/commit/45d58e409e3f51cf0f9a82c1ac8e5a1de79c880a since the refresh method calls `layoutIfNeeded` on the table view, and perhaps there might be a retain cycle somewhere.

</details>

- Ensure that the connectivity is enabled.
- Go to the Reader stream > Discover.
- Quickly switch to different streams.
  - Note: you could also try activating the Network Link Conditioner to switch streams while the request is in progress.
- 🔎 Ensure that the app does not crash.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The card is hidden behind a flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:


Testing checklist:
